syntax = "proto3";

package eventgateway;

import "google/protobuf/timestamp.proto";
import "google/protobuf/struct.proto";
import "google/protobuf/any.proto";

service EventGatewayService {
    // Publish events
    rpc PublishEvent(PublishEventRequest) returns (PublishEventResponse);
    rpc PublishEventsBatch(PublishEventsBatchRequest) returns (PublishEventsBatchResponse);
    
    // Subscribe to events
    rpc Subscribe(SubscribeRequest) returns (stream Event);
    rpc Unsubscribe(UnsubscribeRequest) returns (UnsubscribeResponse);
    
    // Event management
    rpc GetEvent(GetEventRequest) returns (GetEventResponse);
    rpc ListEvents(ListEventsRequest) returns (ListEventsResponse);
    rpc AcknowledgeEvent(AcknowledgeEventRequest) returns (AcknowledgeEventResponse);
    
    // Webhook management
    rpc RegisterWebhook(RegisterWebhookRequest) returns (RegisterWebhookResponse);
    rpc UpdateWebhook(UpdateWebhookRequest) returns (UpdateWebhookResponse);
    rpc DeleteWebhook(DeleteWebhookRequest) returns (DeleteWebhookResponse);
    rpc ListWebhooks(ListWebhooksRequest) returns (ListWebhooksResponse);
    
    // Stream management
    rpc CreateStream(CreateStreamRequest) returns (CreateStreamResponse);
    rpc DeleteStream(DeleteStreamRequest) returns (DeleteStreamResponse);
    rpc ListStreams(ListStreamsRequest) returns (ListStreamsResponse);
}

// CloudEvents message
message CloudEvent {
    string id = 1;
    string source = 2;
    string spec_version = 3;
    string type = 4;
    
    optional string data_content_type = 5;
    optional string data_schema = 6;
    optional string subject = 7;
    optional google.protobuf.Timestamp time = 8;
    
    // Event data
    oneof data {
        google.protobuf.Struct data_json = 9;
        bytes data_binary = 10;
        string data_text = 11;
    }
    
    // Extension attributes
    map<string, string> extensions = 12;
}

message Event {
    CloudEvent cloud_event = 1;
    EventMetadata metadata = 2;
}

message EventMetadata {
    string stream_name = 1;
    uint64 sequence = 2;
    google.protobuf.Timestamp received_at = 3;
    int32 delivery_count = 4;
    map<string, string> headers = 5;
}

message Webhook {
    string id = 1;
    string name = 2;
    string url = 3;
    repeated string event_types = 4;
    
    WebhookConfig config = 5;
    
    google.protobuf.Timestamp created_at = 6;
    google.protobuf.Timestamp updated_at = 7;
    bool enabled = 8;
    
    WebhookStats stats = 9;
}

message WebhookConfig {
    map<string, string> headers = 1;
    string secret = 2;
    int32 timeout_seconds = 3;
    int32 max_retries = 4;
    int32 retry_delay_seconds = 5;
    bool verify_ssl = 6;
}

message WebhookStats {
    int64 total_deliveries = 1;
    int64 successful_deliveries = 2;
    int64 failed_deliveries = 3;
    google.protobuf.Timestamp last_delivery_at = 4;
    string last_delivery_status = 5;
    double average_response_time_ms = 6;
}

message Stream {
    string name = 1;
    string description = 2;
    repeated string subjects = 3;
    StreamConfig config = 4;
    google.protobuf.Timestamp created_at = 5;
}

message StreamConfig {
    int64 max_msgs = 1;
    int64 max_bytes = 2;
    int64 max_age_seconds = 3;
    int32 max_msg_size = 4;
    int32 replicas = 5;
    bool discard_new_per_subject = 6;
}

// Request/Response messages
message PublishEventRequest {
    CloudEvent event = 1;
    string stream = 2;
    map<string, string> headers = 3;
}

message PublishEventResponse {
    string event_id = 1;
    uint64 sequence = 2;
    google.protobuf.Timestamp published_at = 3;
}

message PublishEventsBatchRequest {
    repeated CloudEvent events = 1;
    string stream = 2;
    map<string, string> headers = 3;
}

message PublishEventsBatchResponse {
    repeated PublishEventResponse results = 1;
    int32 succeeded = 2;
    int32 failed = 3;
}

message SubscribeRequest {
    string consumer_id = 1;
    repeated string event_types = 2;
    string stream = 3;
    
    oneof start_position {
        uint64 start_sequence = 4;
        google.protobuf.Timestamp start_time = 5;
        bool deliver_all = 6;
        bool deliver_new = 7;
    }
    
    string durable_name = 8;
    int32 max_in_flight = 9;
    int32 ack_wait_seconds = 10;
}

message UnsubscribeRequest {
    string consumer_id = 1;
    string subscription_id = 2;
}

message UnsubscribeResponse {
    bool success = 1;
    string message = 2;
}

message GetEventRequest {
    string event_id = 1;
}

message GetEventResponse {
    Event event = 1;
}

message ListEventsRequest {
    repeated string event_types = 1;
    string stream = 2;
    google.protobuf.Timestamp start_time = 3;
    google.protobuf.Timestamp end_time = 4;
    int32 limit = 5;
    uint64 start_sequence = 6;
}

message ListEventsResponse {
    repeated Event events = 1;
    bool has_more = 2;
    uint64 next_sequence = 3;
}

message AcknowledgeEventRequest {
    string consumer_id = 1;
    string event_id = 2;
    uint64 sequence = 3;
}

message AcknowledgeEventResponse {
    bool success = 1;
}

message RegisterWebhookRequest {
    Webhook webhook = 1;
}

message RegisterWebhookResponse {
    Webhook webhook = 1;
}

message UpdateWebhookRequest {
    string webhook_id = 1;
    Webhook webhook = 2;
}

message UpdateWebhookResponse {
    Webhook webhook = 1;
}

message DeleteWebhookRequest {
    string webhook_id = 1;
}

message DeleteWebhookResponse {
    bool success = 1;
}

message ListWebhooksRequest {
    repeated string event_types = 1;
    bool enabled_only = 2;
}

message ListWebhooksResponse {
    repeated Webhook webhooks = 1;
}

message CreateStreamRequest {
    Stream stream = 1;
}

message CreateStreamResponse {
    Stream stream = 1;
}

message DeleteStreamRequest {
    string stream_name = 1;
}

message DeleteStreamResponse {
    bool success = 1;
}

message ListStreamsRequest {}

message ListStreamsResponse {
    repeated Stream streams = 1;
}